cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,472 @@
+    boolean onlyMoveImmigrantReplicas = optimizationOptions.onlyMoveImmigrantReplicas();
+    new SortedReplicasHelper().maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(), onlyMoveImmigrantReplicas)
+                              .addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnIncludedTopics(_mustHaveTopicLeadersPerBroker))
+                              .maybeAddPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants(), !onlyMoveImmigrantReplicas)
+                              .trackSortedReplicasFor(_replicaSortName, clusterModel);
+  }
+
+  private int minTopicLeadersPerBroker() {
+    return _balancingConstraint.minTopicLeadersPerBroker();
+  }
+
+  private void validateTopicsWithMinLeaderReplicaIsNotExcluded(OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (optimizationOptions.excludedTopics().isEmpty()) {
+      return;
+    }
+    Set<String> shouldNotBeExcludedTopics = new HashSet<>();
+    _mustHaveTopicLeadersPerBroker.forEach(topicName -> {
+      if (optimizationOptions.excludedTopics().contains(topicName)) {
+        shouldNotBeExcludedTopics.add(topicName);
+      }
+    });
+    if (!shouldNotBeExcludedTopics.isEmpty()) {
+      throw new OptimizationFailureException(String.format("Topics that must have a minimum number of leader replicas "
+          + "per alive broker that is not excluded for leadership moves should not be excluded. This error implies "
+          + "a config error. Topics should not be excluded=[%s] and it is defined as a regex by the config property: %s",
+                                                           String.join(", ", shouldNotBeExcludedTopics),
+                                                           AnalyzerConfig.TOPICS_WITH_MIN_LEADERS_PER_BROKER_CONFIG));
+    }
+  }
+
+  private void validateEnoughLeaderReplicaToDistribute(ClusterModel clusterModel,
+                                                       OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    Map<String, Integer> numLeadersByTopicNames = clusterModel.numLeadersPerTopic(_mustHaveTopicLeadersPerBroker);
+    Set<Broker> eligibleBrokersForLeadership = eligibleBrokersForLeadership(clusterModel, optimizationOptions);
+    int totalMinimumLeaderReplicaCount = eligibleBrokersForLeadership.size() * minTopicLeadersPerBroker();
+
+    for (Map.Entry<String, Integer> numLeadersPerTopic : numLeadersByTopicNames.entrySet()) {
+      if (numLeadersPerTopic.getValue() < totalMinimumLeaderReplicaCount) {
+        throw new OptimizationFailureException(
+            String.format("Cannot distribute %d leader replica(s) over %d broker(s) with total minimum required leader count %d for topic %s.",
+                          numLeadersPerTopic.getValue(), eligibleBrokersForLeadership.size(),
+                          numLeadersPerTopic.getValue(), numLeadersPerTopic.getKey()));
```

Comment:
`total minimum required leader count %d` uses the value `numLeadersPerTopic.getValue()`. But, this uses the same value `numLeadersPerTopic.getValue()`, which was already used in this message.

Should we instead print `minTopicLeadersPerBroker()` with proper message?

Reply:
Yes, good catch!